src/rgw/rgw_op.cc

```diff
@@ -964,6 +965,24 @@ void rgw_build_iam_environment(rgw::sal::Driver* driver,
   }
 }
 
+void handle_replication_status_header(
+    const DoutPrefixProvider *dpp,
+    rgw::sal::Attrs& attrs,
+    rgw::sal::Driver* driver,
+    req_state* s,
+    ceph::real_time &obj_mtime) {
+  auto attr_iter = attrs.find(RGW_ATTR_OBJ_REPLICATION_STATUS);
+  if (attr_iter != attrs.end() && attr_iter->second.to_str() == "PENDING") {
+    RGWObjectSyncHandlerRef sync_handler;
+    int sync_ret = driver->get_object_sync_handler(dpp, &sync_handler, s->yield);
+    if (sync_ret == 0 && sync_handler->object_is_synced(dpp, s->object.get(), obj_mtime)) {
+	attr_iter->second.clear();
+	attr_iter->second.append("COMPLETED");
+	ldpp_dout(dpp, 20) << *s->object << " has amz-replication-status header set to COMPLETED" << dendl;
```

Comment:
do we want to write this change back to the rados head object so we don't have to re-query the log every time?

doing this can be tricky, since that write request could race with another PutObject request to overwrite the object. it looks like `Object::set_obj_attrs()` does the right thing there if you call `Object::set_atomic()` first - `RGWRados::set_attrs()` calls `append_atomic_test()` to verify that the RGW_ATTR_ID_TAG is unchanged